vllm/v1/sample/logits_processor.py

```diff
@@ -0,0 +1,491 @@
+        """Register the removal of a request from
+        the persistent batch.
+
+        Must not be called after the first time
+        self.removed, self.pop_removed_if_can() or
+        self.peek_removed_if_can() are invoked.
+        
+        Args:
+          index: request index
+        """
+        if self._is_removed_sorted:
+            raise RuntimeError("Cannot register new removed request after"
+                               " self.removed has been read.")
+        self._removed.append(index)
+
+    def has_removed(self) -> bool:
+        return bool(self._removed)
+
+    def peek_removed_if_can(self) -> Optional[int]:
+        """Return lowest removed request index"""
+        if self.has_removed():
+            self._sort_removed()
+            return self._removed[-1]
+        return None
+
+    def pop_removed_if_can(self) -> Optional[int]:
+        """Pop lowest removed request index"""
+        if self.has_removed():
+            self._sort_removed()
+            return self._removed.pop()
+        return None
+
+    def reset(self):
+        """Reset batch info at end of step"""
+        self._removed = []
+        self._is_removed_sorted = False
+        self.moved = []
+        self.added = []
+
+    def buildBatchUpdate(self, batch_size: int) -> BatchUpdate:
+        """Generate a logitsprocs batch update data structure.
+        
+        Args:
+          batch_size: current persistent batch size
+
+        Returns:
+          Frozen logitsprocs batch update dataclass instance
+        """
+        return BatchUpdate(
+            batch_size=batch_size,
+            removed=self.removed,
+            moved=self.moved,
+            added=self.added,
+        )
+
+
+class LogitsProcessor(ABC):
+
+    @abstractmethod
+    def apply(self, logits: torch.Tensor) -> torch.Tensor:
+        raise NotImplementedError
+
+    @classmethod
+    @abstractmethod
+    def requires_nongreedy(cls) -> bool:
+        """True if logits processor is incompatible with
+        greedy sampling.
+        TODO(andy): won't be utilized until logits
+        processors are user-extensible
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def update_state(
+        self,
+        batch_update: BatchUpdate,
+    ) -> None:
+        """Called when there are new output tokens, prior
+        to each forward pass.
+
+        Args:
+            batch_update is non-None iff there have been
+            changes to the batch makeup.
+        """
+        raise NotImplementedError
+
+
+@dataclass
+class LogitsProcessorManager:
+    """Encapsulates initialized logitsproc objects.
+
+    Each logits processor has a unique id.
+    """
+    nongreedy: dict[str, LogitsProcessor] = field(
+        default_factory=dict)  # id -> nongreedy-sampling-only logitsproc
+    greedy: dict[str, LogitsProcessor] = field(
+        default_factory=dict)  # id -> greedy-sampling compatible logitsproc
```

Comment:
I think these should be lists (or tuples) rather than dicts.

IIUC the only reason they are dicts is to be able to "look up" min_p for the spec decode check ... we should figure out a different way to handle that.

In the various methods below we are creating and populating new collections each time. These should return iterables and if we need to iterate over all of them we can use `itertools.chain()` (or just have an additional field containing the full list).